src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -314,29 +329,41 @@ protected override void ApplyParseOptionsChanged(ProjectId projectId, ParseOptio
                 throw new ArgumentNullException(nameof(options));
             }
 
-            var parseOptionsService = CurrentSolution.GetProject(projectId).LanguageServices.GetService<IParseOptionsService>();
-            Contract.ThrowIfNull(parseOptionsService, nameof(parseOptionsService));
+            var storage = ProjectPropertyStorage.Create(TryGetDTEProject(projectId), DeferredState.ServiceProvider);
 
-            string newVersion = parseOptionsService.GetLanguageVersion(options);
+            switch (options)
+            {
+                case CSharpCompilationOptions csharpOptions:
+                    storage.SetProperty("AllowUnsafeBlocks", nameof(ProjectConfigurationProperties3.AllowUnsafeBlocks),
+                        csharpOptions.AllowUnsafe.ToString().ToLowerInvariant());
+                    break;
+                case VisualBasicCompilationOptions _:
+                    throw new InvalidOperationException(ServicesVSResources.This_workspace_does_not_support_updating_Visual_Basic_compilation_options);
```

Comment:
shoudl we ave a default case as well?

Reply:
@jasonmalinowski thoughts? just silently ignore, or throw some sort of argument exception?